src/main/java/co/aikar/util/LoadingMap.java

```diff
@@ -0,0 +1,404 @@
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public V get(Object key) {
+        V res = this.backingMap.get(key);
+        if (res == null && key != null) {
+            res = this.loader.apply((K) key);
+            if (res != null) {
+                this.backingMap.put((K) key, res);
+            }
+        }
+        return res;
+    }
+
+    @Override
+    public V put(K key, V value) {
+        return this.backingMap.put(key, value);
+    }
+
+    @Override
+    public V remove(Object key) {
+        return this.backingMap.remove(key);
+    }
+
+    @Override
+    public void putAll(Map<? extends K, ? extends V> m) {
+        this.backingMap.putAll(m);
+    }
+
+    @Override
+    public void clear() {
+        this.backingMap.clear();
+    }
+
+    @Override
+    public Set<K> keySet() {
+        return this.backingMap.keySet();
+    }
+
+    @Override
+    public Collection<V> values() {
+        return this.backingMap.values();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        return this.backingMap.equals(o);
+    }
+
+    @Override
+    public int hashCode() {
+        return this.backingMap.hashCode();
+    }
+
+    @Override
+    public Set<Entry<K, V>> entrySet() {
+        return this.backingMap.entrySet();
+    }
+
+    @Override
+    public LoadingMap<K, V> clone() {
+        return new LoadingMap<>(this.backingMap, this.loader);
+    }
+
+    private static class AutoInstantiatingLoader<K, V> implements Function<K, V> {
+
+        final Constructor<? extends V> constructor;
+        private final Class<? extends V> valueClass;
+
+        AutoInstantiatingLoader(Class<? extends K> keyClass, Class<? extends V> valueClass) {
+            try {
+                this.valueClass = valueClass;
+                if (keyClass != null) {
+                    this.constructor = valueClass.getConstructor(keyClass);
+                } else {
+                    this.constructor = null;
+                }
+            } catch (NoSuchMethodException e) {
+                throw new IllegalStateException(
+                        valueClass.getName() + " does not have a constructor for " + keyClass.getName());
+            }
+        }
+
+        @Override
+        public V apply(K input) {
+            try {
+                return (this.constructor != null ? this.constructor.newInstance(input) : this.valueClass.newInstance());
+            } catch (Exception e) {
+                throw new ExceptionInInitializerError(e);
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return super.hashCode();
+        }
+
+        @Override
+        public boolean equals(Object object) {
+            return false;
+        }
+    }
+
+    /**
+     * Due to java stuff, you will need to cast it to {@link Function}
+     * for some cases.
+     *
+     * @param <T> The value
+     */
+    public static abstract class Feeder<T> implements Function<Object, T> {
```

Comment:
Can you turn this into a interface (FunctionalInterface)? So we can use lamba expressions.